proposals/0075-precompile-for-secp256r1-sigverify.md

```diff
@@ -0,0 +1,298 @@
+taking an opinionated stance on signature malleability, it makes sense to
+also take an opinionated stance on public key encoding.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+The precompile's purpose is to verify signatures using ECDSA-256.
+(denoted in [RFC6460](https://www.ietf.org/rfc/rfc6460.txt) as
+ECDSA using the NIST P-256 curve and the SHA-256 hashing algorithm)
+
+Apart from the RFC mandated implementation the precompile must additionally take
+an opinionated stance on signature malleability.
+
+### Signature Malleability
+
+Due to X axis symmetry along the elliptic curve, for any ECDSA signature
+`(r, s)`, there also exists a valid signature `(r, n - s)`, where `n` is the
+order of the curve. This introduces "s malleability", allowing an attacker
+to produce an alternative version of `s` without invalidating the signature.
+
+The pitfalls of this in authentication systems can be particularly perilous,
+opening up certain implementations to signature replay attacks over the same
+message by simply flipping the `s` value over the curve.
+
+As the primary goal of the `secp256r1` program is secure signature validation
+for authentication purposes, the precompile must mitigate these attacks
+by enforcing the usage of `lowS` values, in which `s <= n/2`.
+
+As such, the program must immediately fail upon the detection of any
+signature that includes a `highS` value. This prevents any accidental
+succeptibility to signature malleability attacks.
+
+Note: The existing `secp256k1` precompile makes no attempt attempt to mitigate
+s malleability, as doing so would go against its primary goal of achieving
+`ecrecover` parity with EVM.
+
+### Implementation
+
+### Program
+
+ID: `Secp256r1SigVerify1111111111111111111111111`
+
+The program instruction must be composed of the following struct:
+
+In Pseudocode:
+
+```
+struct Secp256r1SigVerifyInstruction {
+    count: uint8 LE,                          // Number of signatures to check
+    padding: uint8 LE,                        // Single byte padding
+    signatureOffsets: Array<Secp256r1SignatureOffsets>, // Array of structs
+}
+
+struct Secp256r1SignatureOffsets {
+    signature_offset: uint16 LE,              // Offset to signature
+    signature_instruction_index: uint8 LE,   // Instruction index to signature
+    public_key_offset: uint16 LE,             // Offset to public key
+    public_key_instruction_index: uint8 LE,  // Instruction index to  public key
+    message_data_offset: uint16 LE,           // Offset to start of message data
+    message_data_size: uint16 LE,             // Size of message data
+    message_instruction_index: uint8 LE,     // Instruction index to message
+}
+```
+
+Multiple signatures can be verified. If any of the signatures fail to verify,
+an error must be returned.
+
+If the instruction data is empty, the program must return an error.
+
+If `count == 0` and the length of the instruction data > 1, the program must
+return an error.
```

Comment:
I know existing precompiles have this check, but I think it's unnecessary.

Reply:
So in the case of ```count == 0``` the program passes, irrespective of the instruction data?

Reply:
That would make sense to me, but I'd be interested in hearing from the folks who made this design decision for the other precompiles to make sure I'm not missing something.

Reply:
Will leave this unresolved until we get some more input from the other contributors 🫡

Reply:
For this issue, I think it makes sense either way. This predates me but https://github.com/solana-labs/solana/pull/19300 has the justification for the count, which I think is reasonable.

Reply:
> That would make sense to me, but I'd be interested in hearing from the folks who made this design decision for the other precompiles to make sure I'm not missing something.

This is done as a safeguard for smart-contracts. There have been a number of contracts that wanted to verify a single signature, and thus omitted the check for count.
They just used instruction introspection to assert that the correct data is present in the call to secp verify.

An attacker could thus call secp verify with a count of 0, and pretend to have a valid signature for arbitrary data.

This is obviously a bug in the smart-contract using secp, but having this count=0 check made for an easy fix to prevent this issue from cropping up. It makes no sense to call with a zero count and non-zero data after all.

Reply:
I guess as there's no document or SIMD detailing otherwise, I'd we just stay in line with [#19300](https://github.com/solana-labs/solana/pull/19300).

Thoughts @ptaffet-jump ?